networkx/drawing/layout.py

```diff
@@ -451,117 +525,177 @@ def spring_layout(
+#    except ValueError:
+#        A = nx.to_numpy_array(G, weight=weight)
+#        if k is None and fixed is not None:
+#            # We must adjust k by domain size for layouts not near 1x1
+#            nnodes, _ = A.shape
+#            k = dom_size / np.sqrt(nnodes)
+#        pos = _fruchterman_reingold(
+#            A, k, pos_arr, fixed, iterations, threshold, dim, seed
+#        )
+
+
 fruchterman_reingold_layout = spring_layout
 
 
 @np_random_state(7)
 def _fruchterman_reingold(
-    A, k=None, pos=None, fixed=None, iterations=50, threshold=1e-4, dim=2, seed=None
+    A, k, frame, pos, fixed=None, iterations=50, threshold=1e-4, dim=2, seed=None
 ):
     # Position nodes in adjacency matrix A using Fruchterman-Reingold
     # Entry point for NetworkX graph is fruchterman_reingold_layout()
     import numpy as np
 
+    _, dom_size = frame
+
+    max_extend = np.max(dom_size)
+    t = 0.1 * max_extend
+    dt = t / (iterations + 1)
+
     try:
         nnodes, _ = A.shape
     except AttributeError as err:
         msg = "fruchterman_reingold() takes an adjacency matrix as input"
         raise nx.NetworkXError(msg) from err
 
-    if pos is None:
-        # random initial positions
-        pos = np.asarray(seed.rand(nnodes, dim), dtype=A.dtype)
-    else:
-        # make sure positions are of same type as matrix
-        pos = pos.astype(A.dtype)
-
-    # optimal distance between nodes
-    if k is None:
-        k = np.sqrt(1.0 / nnodes)
-    # the initial "temperature"  is about .1 of domain area (=1x1)
-    # this is the largest step allowed in the dynamics.
-    # We need to calculate this in case our fixed positions force our domain
-    # to be much bigger than 1x1
-    t = max(max(pos.T[0]) - min(pos.T[0]), max(pos.T[1]) - min(pos.T[1])) * 0.1
-    # simple cooling scheme.
-    # linearly step down by dt on each iteration so last iteration is size dt.
-    dt = t / (iterations + 1)
-    delta = np.zeros((pos.shape[0], pos.shape[0], pos.shape[1]), dtype=A.dtype)
     # the inscrutable (but fast) version
     # this is still O(V^2)
     # could use multilevel methods to speed this up significantly
     for iteration in range(iterations):
-        # matrix of difference between points
-        delta = pos[:, np.newaxis, :] - pos[np.newaxis, :, :]
-        # distance between points
-        distance = np.linalg.norm(delta, axis=-1)
-        # enforce minimum distance of 0.01
-        np.clip(distance, 0.01, None, out=distance)
-        # displacement "force"
-        displacement = np.einsum(
-            "ijk,ij->ik", delta, (k * k / distance**2 - A * distance / k)
-        )
-        # update positions
-        length = np.linalg.norm(displacement, axis=-1)
-        length = np.where(length < 0.01, 0.1, length)
-        delta_pos = np.einsum("ij,i->ij", displacement, t / length)
-        if fixed is not None:
-            # don't change positions of fixed nodes
-            delta_pos[fixed] = 0.0
-        pos += delta_pos
```

Comment:
We are not only adding a frame here, we are also adjusting the forces and how the temperature is applied. Thats why I am touching those lines.
I think the `einsum` approach is horrible to read, but if that is the most efficient way to do it, then I can add that again.

Reply:
Yikes -- that sounds like it could change every layout provided... It's a backward compatibility red-flag.
Let me think about what's the best way to handle that. We may need to split the names `spring_layout` and `fr`... But I'm not sure even that will do...

Reply:
Yes, this will change the change the behavior of every layout and I can understand that it's not that easy to handle.
One way to view it is from the perspective of the user who is expecting to run the FR algorithm, but is getting something else. In this case it is not a change to an existing feature, it's a bug-fix.

Simply adding a frame also changes almost every layout, because most of them are large enough to touch the frame.

An other topic is whether these changes actually make things better. I am currently analyzing the outputs of many different layouts, comparing them with the old layout and with what the netgraph package is doing. However I am busy with other things in the next couple of days, so don't expect that to be done this week.